File: backend/src/coordinates.js
const PI = Math.PI;
const EARTH_AXIS_M = 6378245.0;
const ECCENTRICITY_SQUARED = 0.006693421622965943;

function transformLatitude(longitudeOffset, latitudeOffset) {
  let result =
    -100.0 +
    2.0 * longitudeOffset +
    3.0 * latitudeOffset +
    0.2 * latitudeOffset * latitudeOffset +
    0.1 * longitudeOffset * latitudeOffset +
    0.2 * Math.sqrt(Math.abs(longitudeOffset));

  result +=
    ((20.0 * Math.sin(6.0 * longitudeOffset * PI) +
      20.0 * Math.sin(2.0 * longitudeOffset * PI)) *
      2.0) /
    3.0;
  result +=
    ((20.0 * Math.sin(latitudeOffset * PI) +
      40.0 * Math.sin((latitudeOffset / 3.0) * PI)) *
      2.0) /
    3.0;
  result +=
    ((160.0 * Math.sin((latitudeOffset / 12.0) * PI) +
      320.0 * Math.sin((latitudeOffset * PI) / 30.0)) *
      2.0) /
    3.0;
  return result;
}

function transformLongitude(longitudeOffset, latitudeOffset) {
  let result =
    300.0 +
    longitudeOffset +
    2.0 * latitudeOffset +
    0.1 * longitudeOffset * longitudeOffset +
    0.1 * longitudeOffset * latitudeOffset +
    0.1 * Math.sqrt(Math.abs(longitudeOffset));

  result +=
    ((20.0 * Math.sin(6.0 * longitudeOffset * PI) +
      20.0 * Math.sin(2.0 * longitudeOffset * PI)) *
      2.0) /
    3.0;
  result +=
    ((20.0 * Math.sin(longitudeOffset * PI) +
      40.0 * Math.sin((longitudeOffset / 3.0) * PI)) *
      2.0) /
    3.0;
  result +=
    ((150.0 * Math.sin((longitudeOffset / 12.0) * PI) +
      300.0 * Math.sin((longitudeOffset / 30.0) * PI)) *
      2.0) /
    3.0;
  return result;
}

export function isOutsideGcj02Coverage(longitudeDeg, latitudeDeg) {
  return (
    longitudeDeg < 72.004 ||
    longitudeDeg > 137.8347 ||
    latitudeDeg < 0.8293 ||
    latitudeDeg > 55.8271
  );
}

function roundedCoordinate(value) {
  return Number(value.toFixed(7));
}

/**
 * Converts a validated WGS84 point to the coordinate system expected by AMap.
 * Points outside the mainland-China GCJ-02 coverage bounds are left unchanged.
 */
export function wgs84ToGcj02(point) {
  const longitudeDeg = Number(point.longitude_deg);
  const latitudeDeg = Number(point.latitude_deg);

  if (!Number.isFinite(longitudeDeg) || !Number.isFinite(latitudeDeg)) {
    throw new TypeError("WGS84 point must contain finite longitude_deg and latitude_deg");
  }

  if (isOutsideGcj02Coverage(longitudeDeg, latitudeDeg)) {
    return {
      coordinate_system: "GCJ-02",
      longitude_deg: roundedCoordinate(longitudeDeg),
      latitude_deg: roundedCoordinate(latitudeDeg),
    };
  }

  let latitudeDelta = transformLatitude(longitudeDeg - 105.0, latitudeDeg - 35.0);
  let longitudeDelta = transformLongitude(longitudeDeg - 105.0, latitudeDeg - 35.0);
  const radianLatitude = (latitudeDeg / 180.0) * PI;
  let magic = Math.sin(radianLatitude);
  magic = 1 - ECCENTRICITY_SQUARED * magic * magic;
  const sqrtMagic = Math.sqrt(magic);
  latitudeDelta =
    (latitudeDelta * 180.0) /
    (((EARTH_AXIS_M * (1 - ECCENTRICITY_SQUARED)) / (magic * sqrtMagic)) * PI);
  longitudeDelta =
    (longitudeDelta * 180.0) /
    ((EARTH_AXIS_M / sqrtMagic) * Math.cos(radianLatitude) * PI);

  return {
    coordinate_system: "GCJ-02",
    longitude_deg: roundedCoordinate(longitudeDeg + longitudeDelta),
    latitude_deg: roundedCoordinate(latitudeDeg + latitudeDelta),
  };
}

/**
 * Converts an AMap GCJ-02 point back to the WGS84 contract used by GNSS
 * clients. GCJ-02 has no closed-form inverse, so this uses a bounded binary
 * search around the source point and verifies every candidate through the
 * same forward transform used for route requests.
 */
export function gcj02ToWgs84(point) {
  const longitudeDeg = Number(point.longitude_deg);
  const latitudeDeg = Number(point.latitude_deg);

  if (!Number.isFinite(longitudeDeg) || !Number.isFinite(latitudeDeg)) {
    throw new TypeError("GCJ-02 point must contain finite longitude_deg and latitude_deg");
  }

  if (isOutsideGcj02Coverage(longitudeDeg, latitudeDeg)) {
    return {
      coordinate_system: "WGS84",
      longitude_deg: roundedCoordinate(longitudeDeg),
      latitude_deg: roundedCoordinate(latitudeDeg),
    };
  }

  let minimumLongitude = longitudeDeg - 0.02;
  let maximumLongitude = longitudeDeg + 0.02;
  let minimumLatitude = latitudeDeg - 0.02;
  let maximumLatitude = latitudeDeg + 0.02;
  let candidateLongitude = longitudeDeg;
  let candidateLatitude = latitudeDeg;

  for (let iteration = 0; iteration < 32; iteration += 1) {
    candidateLongitude = (minimumLongitude + maximumLongitude) / 2;
    candidateLatitude = (minimumLatitude + maximumLatitude) / 2;
    const projected = wgs84ToGcj02({
      longitude_deg: candidateLongitude,
      latitude_deg: candidateLatitude,
    });

    if (projected.longitude_deg < longitudeDeg) {
      minimumLongitude = candidateLongitude;
    } else {
      maximumLongitude = candidateLongitude;
    }
    if (projected.latitude_deg < latitudeDeg) {
      minimumLatitude = candidateLatitude;
    } else {
      maximumLatitude = candidateLatitude;
    }
  }

  return {
    coordinate_system: "WGS84",
    longitude_deg: roundedCoordinate(candidateLongitude),
    latitude_deg: roundedCoordinate(candidateLatitude),
  };
}

File: shared/coordinates/coordinates.cpp
#include "moto_coordinates.hpp"

#include <cmath>

namespace moto::coordinates {
namespace {

constexpr double kPi = 3.141592653589793238462643383279502884;
constexpr double kEarthAxisM = 6378245.0;
constexpr double kEccentricitySquared = 0.006693421622965943;
constexpr double kCoordinateScale = 10000000.0;

double transform_latitude(double longitude_offset,
                          double latitude_offset) noexcept {
  double result = -100.0 + 2.0 * longitude_offset + 3.0 * latitude_offset +
                  0.2 * latitude_offset * latitude_offset +
                  0.1 * longitude_offset * latitude_offset +
                  0.2 * std::sqrt(std::fabs(longitude_offset));

  result += ((20.0 * std::sin(6.0 * longitude_offset * kPi) +
              20.0 * std::sin(2.0 * longitude_offset * kPi)) *
             2.0) /
            3.0;
  result += ((20.0 * std::sin(latitude_offset * kPi) +
              40.0 * std::sin((latitude_offset / 3.0) * kPi)) *
             2.0) /
            3.0;
  result += ((160.0 * std::sin((latitude_offset / 12.0) * kPi) +
              320.0 * std::sin((latitude_offset * kPi) / 30.0)) *
             2.0) /
            3.0;
  return result;
}

double transform_longitude(double longitude_offset,
                           double latitude_offset) noexcept {
  double result = 300.0 + longitude_offset + 2.0 * latitude_offset +
                  0.1 * longitude_offset * longitude_offset +
                  0.1 * longitude_offset * latitude_offset +
                  0.1 * std::sqrt(std::fabs(longitude_offset));

  result += ((20.0 * std::sin(6.0 * longitude_offset * kPi) +
              20.0 * std::sin(2.0 * longitude_offset * kPi)) *
             2.0) /
            3.0;
  result += ((20.0 * std::sin(longitude_offset * kPi) +
              40.0 * std::sin((longitude_offset / 3.0) * kPi)) *
             2.0) /
            3.0;
  result += ((150.0 * std::sin((longitude_offset / 12.0) * kPi) +
              300.0 * std::sin((longitude_offset / 30.0) * kPi)) *
             2.0) /
            3.0;
  return result;
}

double round_coordinate(double value) noexcept {
  return std::round(value * kCoordinateScale) / kCoordinateScale;
}

}  // namespace

bool is_valid_coordinate(Coordinate coordinate) noexcept {
  return std::isfinite(coordinate.longitude_deg) &&
         std::isfinite(coordinate.latitude_deg);
}

bool is_outside_gcj02_coverage(Coordinate coordinate) noexcept {
  return coordinate.longitude_deg < 72.004 ||
         coordinate.longitude_deg > 137.8347 ||
         coordinate.latitude_deg < 0.8293 ||
         coordinate.latitude_deg > 55.8271;
}

ConversionResult wgs84_to_gcj02(Coordinate coordinate) noexcept {
  if (!is_valid_coordinate(coordinate)) {
    return {coordinate, ConversionStatus::InvalidInput};
  }

  if (is_outside_gcj02_coverage(coordinate)) {
    return {{round_coordinate(coordinate.longitude_deg),
             round_coordinate(coordinate.latitude_deg)},
            ConversionStatus::Ok};
  }

  double latitude_delta =
      transform_latitude(coordinate.longitude_deg - 105.0,
                         coordinate.latitude_deg - 35.0);
  double longitude_delta =
      transform_longitude(coordinate.longitude_deg - 105.0,
                          coordinate.latitude_deg - 35.0);
  const double radian_latitude = coordinate.latitude_deg / 180.0 * kPi;
  double magic = std::sin(radian_latitude);
  magic = 1.0 - kEccentricitySquared * magic * magic;
  const double sqrt_magic = std::sqrt(magic);
  latitude_delta =
      latitude_delta * 180.0 /
      ((kEarthAxisM * (1.0 - kEccentricitySquared) /
        (magic * sqrt_magic)) *
       kPi);
  longitude_delta = longitude_delta * 180.0 /
                    ((kEarthAxisM / sqrt_magic) *
                     std::cos(radian_latitude) * kPi);

  return {{round_coordinate(coordinate.longitude_deg + longitude_delta),
           round_coordinate(coordinate.latitude_deg + latitude_delta)},
          ConversionStatus::Ok};
}

}  // namespace moto::coordinates
